applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -187,28 +189,64 @@ void data_send(const uint8_t *data, size_t len)
 	}
 }
 
-static int uart_receive(void)
+static int uart_receive_enable(void)
 {
 	int ret;
 
+	if (k_sem_take(&rx_change, K_MSEC(UART_RX_CHANGE_TIMEOUT_MS)) != 0) {
+		LOG_ERR("UART RX failed to enable");
+		(void)uart_send(FATAL_STR, sizeof(FATAL_STR) - 1);
+		return -EAGAIN;
+	}
+
+	next_buf = uart_rx_buf[1];
 	ret = uart_rx_enable(uart_dev, uart_rx_buf[0], sizeof(uart_rx_buf[0]), UART_RX_TIMEOUT_US);
 	if (ret && ret != -EBUSY) {
-		LOG_ERR("UART RX failed: %d", ret);
+		LOG_ERR("UART RX enable failed: %d", ret);
 		(void)uart_send(FATAL_STR, sizeof(FATAL_STR) - 1);
+		k_sem_give(&rx_change);
 		return ret;
 	}
-	next_buf = uart_rx_buf[1];
 	at_buf_overflow = false;
 	at_buf_len = 0;
 
+	k_sem_give(&rx_change);
+
 	return 0;
 }
 
+static void uart_receive_disable(bool blocking)
```

Comment:
Discussed with @junqingzou and this should not be called blocking, but for example wait_for_disable.